### PR TITLE
ivy: statically distinguish local from global variables

### PR DESCRIPTION
**Pull request by rsc**
Following Python's lead, a local variable is one that is written
before being read in the program execution order, and
a global variable is one that is read before being written.

The statement 'x;' serves to declare x as a global, assuming
x has not been assigned yet. We could always define an operator too.

Because we have this analysis, we can also count the number
of local variables, assign them indices, and use a slice instead of a map
to hold them, so this change does that as well.

FOR DISCUSSION. It is unclear whether this is the right rule,
and probably there is still polishing to do. It is unclear where
doc changes should be applied, for example.

### CODE REVIEW

**Comment by rsc** [comment]
This still needs docs, and the exact code details will depend on the fix for #95, because the read/write checker should match the order of evaluation at run time. But here it is as a PR for now.

Where would be a good place to document this behavior?

**Comment by robpike** [comment]
I landed the fix to #95, which was trivial. Documentation for the behavior here should go in doc.go in the section on ops (")help ops"). The final paragraph needs an update - it even mentions a possible change this CL addresses.

File: exec/context.go [hunk 1]
```diff
@@ -22,9 +22,11 @@ type Context struct {
 	// Accessed through the value.Context Config method.
 	config *config.Config
 
-	// Stack is a stack of symbol tables, one entry per function (op) invocation,
-	// plus the 0th one at the base.
-	Stack []Symtab
+	frames []int // number of values in each stack frame
```
Comment (robpike) [inline exec/context.go:25]: maybe frameSize? "frames" is a funny name for a count.

File: exec/context.go [hunk 2]
```diff
@@ -56,55 +58,48 @@ func (c *Context) Config() *config.Config {
...
-	syms["e"], syms["pi"] = value.Consts(c)
+	e, pi := value.Consts(c)
+	c.AssignGlobal("e", e)
+	c.AssignGlobal("pi", pi)
 }
 
-// Lookup returns the value of a symbol.
-func (c *Context) Lookup(name string) value.Value {
-	return c.varFrame(name, false)[name]
+// Global returns the value of a global symbol.
```
Comment (robpike) [inline exec/context.go:66]: or nil if the symbol is not defined globally.

File: testdata/statement.ivy [hunk 3]
```diff
@@ -159,3 +159,15 @@ x = 4 4 rho 0; y = 4 4; x[y[1]; y[2]] = 3; x
 	0 0 0 0
 	0 0 0 0
 	0 0 0 3
+
+# Order of evaluation
```
Comment (robpike) [inline testdata/statement.ivy:163]: i was going to add this to my fix for #95, but left it for you. i did need to fix one of the existing examples, demonstrating that the fix matters.

**Review by robpike — CHANGES_REQUESTED** [review]

**Comment by rsc** [comment]
Applied all suggestions, added docs.

**Review by robpike — APPROVED** [review]

**Pull request merged**